### PR TITLE
enh(php) Switch highlighter to partially case-insensitive

### PR DESCRIPTION
**Pull request by wkania**
### Changes
After [discussion](https://github.com/highlightjs/highlight.js/pull/3422#discussion_r766006060) related to PHP partial case-insensitive .
- [x] Numbers ([case-insensitive](https://www.php.net/manual/en/language.types.integer.php))
- [x] Variables ([case-sensitive](https://www.php.net/manual/en/language.variables.basics.php))
- [x] class names (case-insensitive)
- [x] function names (case-insensitive)
- [x] keywords (both)

### Checklist
- [x] Added markup tests
- [x] Updated the changelog at `CHANGES.md`

### CODE REVIEW

**Comment by joshgoebel** [comment]
Actually let me rewrite the keyword lists to arrays first so the PR will be easier to follow... then you would just need a `bothCases` helper function that would take an array and spit out another array with both variants, ie:

- ["then"]  => ["then", "THEN"]
- ["__FILE__"]  => ["__FILE__", "__file__"]

Assuming the magic constants are case-insensitive.

**Comment by joshgoebel** [comment]
https://github.com/highlightjs/highlight.js/pull/3427

**Comment by wkania** [comment]
> Actually let me rewrite the keyword lists to arrays first so the PR will be easier to follow... then you would just need a `bothCases` helper function that would take an array and spit out another array with both variants, ie:
> 
>     * ["then"]  => ["then", "THEN"]
> 
>     * ["**FILE**"]  => ["**FILE**", "**file**"]
> 
> 
> Assuming the magic constants are case-insensitive.

Great but there are more valid cases like:
 ` If(true) { };   
  iF(true) { };`
  
  Is there an option to overwrite global case sensitivity with the `match: /if/i` ?

**Comment by joshgoebel** [comment]
> Great but there are more valid cases like: If(true) { }; iF(true) { };

But do people do that? That seems very UGLY/bad practice to me. (esp the latter)

We don't have any easy way to do that... all the grammar regex must be `/i` or not `/i`, they can NOT be mixed, the sensitivity is controlled by the setting as I mentioned earlier.

`keywords` is done by raw string matching so it's easy to hack for `BLAH` and `blah`.  We could possibly also calculate a "Blah", but I probably would NOT do it by default... I'd wait and see how common that was in the wild.

**Comment by wkania** [comment]
> But do people do that? That seems very UGLY/bad practice to me. (esp the latter)

People rather use lowercase. I saw some cases of uppercase: TRUE, FALSE, NULL. So we can ignore it.

**Comment by joshgoebel** [comment]
Oh I wonder if it's more common for literals, we could just support those if those are common...

**Comment by joshgoebel** [comment]
Can you give me commit access to your fork?

**Comment by joshgoebel** [comment]
Do you think we could match function dispatch similar to how we do in CPP so we just catch all invocations?

https://github.com/highlightjs/highlight.js/blob/main/src/languages/cpp.js#L401

**Comment by wkania** [comment]
> Do you think we could match function dispatch similar to how we do in CPP so we just catch all invocations?
> 
> https://github.com/highlightjs/highlight.js/blob/main/src/languages/cpp.js#L401

Looks good. Invocations of?

**Comment by joshgoebel** [comment]
> Invocations of?

Any function... if it's a function we can highlight it as `title.function`...

**Comment by wkania** [comment]
> > Invocations of?
> 
> Any function... if it's a function we can highlight it as `title.function`...

That what I thought. Currently, the highlighter does not support function invocation, only declarations. Do you want this to be part of this MR?

**Comment by joshgoebel** [comment]
Well you had "function names (case-insensitive)" on your list... so not that's the first thing I jumped too... overall we're trying to move away from complex parsing an do simpler pattern matching... so for many languages the same rule can be used to detect function definition and function invocation.

```
function blah() ...
blah(32)
```

The same `[ident](` rule hits both.

**Comment by joshgoebel** [comment]
If 99% of the time people use lower case keywords I'm ok with just trying that and seeing how it works (with the exception of the literals)... would that complete `keywords (both)` then?

**Comment by wkania** [comment]
> If 99% of the time people use lower case keywords I'm ok with just trying that and seeing how it works (with the exception of the literals)... would that complete `keywords (both)` then?

This should be fine. Will look at it tomorrow.

**Comment by wkania** [comment]
@joshgoebel Ready for review. 
Keywords match|0 and Error|0 were problematic.

**Comment by joshgoebel** [comment]
> Keywords match|0 and Error|0 were problematic.

How so?

File: src/languages/php.js [hunk 1]
```diff
@@ -263,13 +262,65 @@ export default function(hljs) {
...
+      }
+    });
+    return result;
+  };
+
   const KEYWORDS = {
-    keyword: KWS,
-    literal: LITERALS,
-    built_in: BUILT_INS
+    keyword: KWS.concat([ "match|0" ]),
```
Comment (joshgoebel) [inline src/languages/php.js:286]: No, this is not the right solution.  The keyword lists should be what they are.  If the `|0` tagging causes a problem for your code on line 312 (or elsewhere) then you need to write a filter function that returns just raw keywords by stripping the `|0`...

File: src/languages/php.js [hunk 2]
```diff
@@ -263,13 +262,65 @@ export default function(hljs) {
...
+        },
+      }
+    ]
+  };
+
+  const FUNCTION_INVOKE = {
+    relevance: 0,
+    match: [
+      /(?:->|::|\s|\(|\\)/,
+      regex.concat("(?!fn\\b|function\\b|match\\b|Error\\b|", KWS.join("\\b|"), "|", BUILT_INS.join("\\b|"), "\\b)"),
```
Comment (joshgoebel) [inline src/languages/php.js:312]: How many of the builtins and keywords are truly used like functions?  Is it a majority?  In most languages it's only a few... if there was a list of 5-10 in PHP I'd rather be explicit rather than just negative match the ENTIRE world of keywords (which does have a cost).

Comment (joshgoebel) [reply src/languages/php.js:312]: Or if we just added an explicit rule for `[BUILT_IN](` (if many are used like functions) that would also remove the need for all the negative matching of built-ins here...

File: src/languages/php.js (hunk 1 above)
Comment (joshgoebel) [reply src/languages/php.js:286]: But as I mention elsewhere I also question the need to use the full lists as a negative filter...

File: src/languages/php.js (hunk 2 above)
Comment (joshgoebel) [reply src/languages/php.js:312]: Also if you are doing negative filters then  you need a few markup tests to test the negative case as well to show that your negative filter is working as intended (you can add comments to explain the purposes of the negative tests).

Comment (wkania) [reply src/languages/php.js:312]: > How many of the builtins and keywords are truly used like functions? Is it a majority? In most languages it's only a few... if there was a list of 5-10 in PHP I'd rather be explicit rather than just negative match the ENTIRE world of keywords (which does have a cost).




```
  const MIGHT_BE_BEFORE_OPENING_BRACKET = [
    "array",
    "case",
    "catch",
    "die",
    "echo",
    "exit",
    "include",
    "include_once",
    "print",
    "require",
    "require_once",
    "do",
    "else",
    "elseif",
    "empty",
    "eval",
    "finally",
    "for",
    "foreach",
    "fn",
    "function",
    "if",
    "int",
    "isset",
    "list",
    "match",
    "return",
    "switch",
    "throw",
    "trait",
    "try",
    "unset",
    "use",
    "while",
  ];
```
This was my first approach ^.
There will be also some from `BUILT_INS`.



> Also if you are doing negative filters then you need a few markup tests to test the negative case as well to show that your negative filter is working as intended (you can add comments to explain the purposes of the negative tests).

I reused old tests. New code didn't break them, but some comment would be in place. I will add some extra markup tests

**Comment by joshgoebel** [comment]
> This was my first approach ^.

Ugh, ok - that's a lot... :-) We can keep it as-is and see if it works ok I suppose.

File: src/languages/php.js (hunk 2 above)
Comment (wkania) [reply src/languages/php.js]: > Also if you are doing negative filters then you need a few markup tests to test the negative case as well to show that your negative filter is working as intended (you can add comments to explain the purposes of the negative tests).

 @joshgoebel Added extra tests and some comments.

File: src/languages/php.js [hunk 3]
```diff
@@ -263,13 +264,80 @@ export default function(hljs) {
...
+  /**
+   * @param {string[]} items */
+  const normalizeKeywords = (items) => {
+    /** @type string[] */
+    const result = [];
+    items.forEach(item => {
+      item.replace(/\|0$/, "");
+      result.push(item);
+    });
+    return result;
```
Comment (joshgoebel) [inline src/languages/php.js:302]: Use `map` for transforms?

Comment (joshgoebel) [reply src/languages/php.js:302]: Also this needs to match `\d+` not just `0`... so it's future proof.

Comment (joshgoebel) [reply src/languages/php.js:302]: I think this is all I see.  I might step in after and make a negative look-ahead list helper... but this looks pretty clean right now.

Comment (wkania) [reply src/languages/php.js]: @joshgoebel done

> I think this is all I see. I might step in after and make a negative look-ahead list helper... but this looks pretty clean right now.

Good idea

File: src/languages/php.js [hunk 4]
```diff
@@ -263,13 +264,76 @@ export default function(hljs) {
...
+          4: "title.class",
+        },
+      }
+    ]
   };
+
+  const FUNCTION_INVOKE = {
+    relevance: 0,
+    match: [
+      /(?:->|::|\s|\(|\\)/,
```
Comment (joshgoebel) [inline src/languages/php.js:323]: Please provide commented examples of all the variants you're using here. Is just an attempt to eliminate false positives - if so what would an example or two be?   Why do we need these at all vs just relying on `[ident](`...?

Comment (wkania) [reply src/languages/php.js:324]: > Is just an attempt to eliminate false positives - if so what would an example or two be? Why do we need these at all vs just relying on `[ident](`...?
> 
> If this is critical please provide commented code examples of all the variants you're using here.

All the correct cases are in the `Function invoke` tests. I also added a comment and test for `First-class Callable Syntax`,
which was working correctly before this PR (because there was no function invoke). Now it would be a false positive, so I added all the rules.

Comment (joshgoebel) [reply src/languages/php.js:324]: I'm not following...   Isn't this just an issue with the non-function guard?  If I do this, all tests pass:

```
diff --git a/src/languages/php.js b/src/languages/php.js
index 9f45841a..297bd55b 100644
--- a/src/languages/php.js
+++ b/src/languages/php.js
@@ -321,7 +321,8 @@ export default function(hljs) {
     relevance: 0,
     match: [
       // to prevent First-class Callable Syntax from being confused as the function title
-      /(?:->|::|\s|\(|\\)/,
+      // /(?:->|::|\s|\(|\\)/,
+      /\b/,
       // to prevent keywords from being confused as the function title
       regex.concat("(?!fn\\b|function\\b|", normalizeKeywords(KWS).join("\\b|"), "|", normalizeKeywords(BUILT_INS).join("\\b|"), "\\b)"),
       /\w+/,
@@ -375,7 +376,8 @@ export default function(hljs) {
       FUNCTION_INVOKE,
       {
         // swallow composed identifiers to avoid parsing them as keywords
-        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*/
+        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?!\s*\()(?![a-zA-Z0-9_\x7f-\xff])/,
+        // scope:"wrong"
       },
       CONSTRUCTOR_CALL,
       {
```

IE, fixing the guard and using only a word boundary to match the start of function invokes... thoughts?  This using multi-match as "look-behind" is dangerous and I really don't want to encourage it at all if it can be avoided.

Comment (wkania) [reply src/languages/php.js:324]: @joshgoebel Strange because I applied your fix and they fall.
```
1 failing

  1) php
       should not cause polynomial backtracking:
     Error: php/[10]/begin: Polynomial backtracking. By repeating any character that matches /\xa0/i, an attack string can be created.

    [a-zA-Z0-9_\x7f-\xff]*(?!\s*
    ^~~~~~~~~~~~~~~~~~~~~~   ^~~

Full pattern:
/(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?!\s*\()(?![a-zA-Z0-9_\x7f-\xff])/m

```

After changing `\s*` to `\s` I have a below error and a few offers.

```
-DateTimeImmutable::createFromMutable(<span class="hljs-keyword">new</span> <span class="hljs-title class_">\DateTime</span>(<span class="hljs-string">&#x27;now&#x27;</span>));
      +DateTimeImmutable::<span class="hljs-function title_ invoke__">createFromMutable</span>(<span class="hljs-keyword">new</span> <span class="hljs-title class_">\DateTime</span>(<span class="hljs-string">&#x27;now&#x27;</span>));
```
This regex do not detect static function invoke:
`DateTimeImmutable::createFromMutable();`
or method invoke
`$date->format();`
and we do not want to higlight First-class Callable Syntax as a function
`$test();`

P.S. I'm run tests inside Docker.

Comment (joshgoebel) [reply src/languages/php.js:324]: > should not cause polynomial backtracking:

That can be ignored [for now] and fixed.  **I didn't have any markup test errors.**  Your'e saying you're seeing the same result right - all tests pass , other than the security issue?

So if we can fix that, then can this be simplified as I'm suggesting?

Comment (joshgoebel) [reply src/languages/php.js:324]: Can names TRULY contain non-breaking spaces (0xA0)? That sounds like a bug to me...

Comment (wkania) [reply src/languages/php.js:324]: > > should not cause polynomial backtracking:
> 
> That can be ignored [for now] and fixed. **I didn't have any markup test errors.** Your'e saying you're seeing the same result right - all tests pass with my patch , other than the security issue?
> 
> So if we can fix that, then can this be simplified as I'm suggesting?

It will be annoying



> Can names TRULY contain non-breaking spaces (0xA0)? That sounds like a bug to me...

[Yes](https://mnapoli.fr/using-non-breakable-spaces-in-test-method-names/), but I never saw anyone using it. Looks bad for me.

Comment (wkania) [reply src/languages/php.js:324]: @joshgoebel fixed it with ` *` because only space is valid before the name of the function and opening bracket.

File: src/languages/php.js [hunk 5]
```diff
@@ -374,7 +374,8 @@ export default function(hljs) {
       FUNCTION_INVOKE,
       {
         // swallow composed identifiers to avoid parsing them as keywords
-        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*/
+        begin: /(::|->)+[a-zA-Z_\x7f-\xff][a-zA-Z0-9_\x7f-\xff]*(?! *\()(?![a-zA-Z0-9_\x7f-\xff])/,
```
[the remark this answers is not in the document]
Comment (wkania) [reply src/languages/php.js:377]: > _I'm not sure they all needed to change_, unless the `[ ]` is more precise? (it may likely be) IIRC the issue was with this line alone because `\xA0*` overlaps with `\s*`.... which is fixed easily enough by changing the look-ahead \s to a space, if that works for our use cases. :-) The character groups in the regex here also could have been fixed. (I might just do this)
> 
> So, is this more precise? (in which case I'll likely change them to `[ ]` for readability.
> 
> Any final thoughts? Thanks for all the hard work on this. I think the finish line is just ahead. :)

```
new    
\DateTime   
();
```
This is also allowed so I will introduce const SEPARATOR `[ \t\n]`. This should cover all the cases. Probably this will be better in all those places, if non-breaking space may be part of the name and `\s` select it.

[the remark this answers is not in the document]
Comment (joshgoebel) [reply src/languages/php.js:377]: > This is also allowed 

I imagined it might be. :-)

> so I will introduce const SEPARATOR

I think "WHITESPACE" might be better (and add a comment saying it purposely excludes `0xA0`), etc...

[the remark this answers is not in the document]
Comment (wkania) [reply src/languages/php.js]: > I think "WHITESPACE" might be better (and add a comment saying it purposely excludes `0xA0`), etc...

Done @joshgoebel

**Review by joshgoebel — APPROVED** [review]

**Comment by joshgoebel** [comment]
@wkania Thanks for all the effort.  I might release 11.4 soon just to see how this case change plays out.
